content/blog/2019-11-13-space-efficient-gc/index.md

```diff
@@ -0,0 +1,100 @@
+"""
+latex = true
+
++++
+
+["Space efficient conservative garbage collection"](https://dl.acm.org/citation.cfm?id=155109) describes some inexpensive but useful techniques that can make *conservative garbage collectors* more space efficient.
+Their techniques can reduce pointer misidentification to retain less memory during runtime, and also excess retention due to less information about variable liveness. Their methods can be easily incorporated into any garbage collecting allocator transparently to client programs.
+
+
+# Pointer Misidentification and Solutions od Existing Conservative Garbage Collectors
+
+The goal of a garbage collector is to retain as little memory as it can, subject to the constraint that all memory that will be accessed in the future must be retained. A garbage collector is called **conservative** if it can operate with minimal information about the layout of the client program's data. 
+
+For conservative collectors, the most apparent potential source of excess memory retention is **pointer misidentification** (e.g., misidentify integers as pointers). Here are some examples:
+
+- An integer variable contains the address of a valid but inaccessible object
+- An object that has a pointer pointing to the interior
+
+The probability of such misidentification even get increased if more of the address space is occupied by the heap.
+
+Some *current ad hoc techniques* can help decrease pointer misidentification probablity. One method is to design an allocator to avoid allocating objects at address that are likely to collide with other data (properly position the heap in the address space) and align pointers properly. Otherwise, all possible alignments must be considered by the collector and could result in more false pointers. For example, in the figure shown below, these two small integers (*0000 0009* and *0000 000a*) can also be viewed as a valid heap address (*0009 0000*) with the concatenation of the low order half word of the first integer and the high order half of the next.
+
+![001](./001.png)
+
+Currently, most compilers always guarantee adequate alignment. However, it is hard to determine the proper position of the heap. Thus, this paper introduces a much less ad hoc and more flexible technique to avoid pointer misidentification.
+
+
+# Their Techniques
+
+This paper's techniques has two steps: 
+
+### Regular Garbage Collection
+
+They ensure regular garbage collections, with at least a fast and initial one happening just after system startup before any allocation begins.
+
+### Invalid Pointer Record
+
+A record of invalid pointers found during a garbage collection that could be valid object addresses afterwards is kept as a **blacklist**. The figure below shows the modified marking algorithm with blacklisting. The only modification is in bold, which adds adjacent addresses to the blacklist.
```

Comment:
Can you clarify how the system knows that a value is an invalid pointer? Without a little more detail, it sounds a little cyclic: the GC needs to know which pointers are invalid, and to do so, it checks whether the pointer is invalid.